Movecraft/src/main/java/net/countercraft/movecraft/processing/tasks/detection/DetectionTask.java

```diff
@@ -339,35 +352,79 @@ private DetectAction(ConcurrentLinkedQueue<MovecraftLocation> currentFrontier, C
+                    BlockFace facing = toBlockFace(lantern);
+                    MovecraftLocation relativeLoc = probe.getRelative(facing);
+
+                    if (!legal.contains(relativeLoc)) {
+                        blockFacingCraft = false; // Invalidate block if it's not facing the craft
+                    }
+                }
+
+                if (!blockFacingCraft) {
+                    Bukkit.getLogger().info("Skipped");
+                    continue;
+                }
+
+                if(!visited.add(probe))
+                    continue;
+
+                visitedMaterials.computeIfAbsent(m, Functions.forSupplier(ConcurrentLinkedDeque::new)).add(probe);
                 if(!ALLOWED_BLOCK_VALIDATOR.validate(probe, type, movecraftWorld, player).isSucess())
                     continue;
 
-                DetectionPredicate<MovecraftLocation> chain = FORBIDDEN_BLOCK_VALIDATOR;
-                for(var validator : VALIDATORS) {
-                    chain = chain.and(validator);
-                }
                 var result = chain.validate(probe, type, movecraftWorld, player);
 
-                if(result.isSucess()) {
-                    legal.add(probe);
-                    if(Tags.FLUID.contains(movecraftWorld.getMaterial(probe)))
-                        fluid.add(probe);
-
-                    size.increment();
-                    materials.computeIfAbsent(movecraftWorld.getMaterial(probe), Functions.forSupplier(ConcurrentLinkedDeque::new)).add(probe);
-                    for(MovecraftLocation shift : SHIFTS) {
-                        var shifted = probe.add(shift);
-                        if(visited.add(shifted))
-                            nextFrontier.add(shifted);
-                    }
-                }
-                else {
+                if (!result.isSucess()) {
                     illegal.add(probe);
                     audience.sendMessage(Component.text(result.getMessage()));
+                    return;
+                }
+
+                legal.add(probe);
+                if (Tags.FLUID.contains(m))
+                    fluid.add(probe);
+
+                size.increment();
+                materials.computeIfAbsent(m, Functions.forSupplier(ConcurrentLinkedDeque::new)).add(probe);
+                for (MovecraftLocation shift : SHIFTS) {
+                    var shifted = probe.add(shift);
+                    nextFrontier.add(shifted);
                 }
             }
         }
+
+        private BlockFace toBlockFace(FaceAttachable.AttachedFace attachedFace, BlockData bd) {
+            return switch (attachedFace) {
+                case FLOOR -> BlockFace.DOWN;
+                case WALL -> ((Directional) bd).getFacing().getOppositeFace();
+                case CEILING -> BlockFace.UP;
+            };
+        }
+
+        private BlockFace toBlockFace(Lantern hangable) {
+            return hangable.isHanging() ? BlockFace.UP : BlockFace.DOWN;
+        }
```

Comment:
I'd like to see an `Optional<BlockFace> getSupportFace(BlockData data)` static method that exists somewhere outside of this class to simplify the implementation logic. The method should match against as many block types that require support as possible, and leave it to the user to configure which blocks are actually checked.

Reply:
Should I put that into `MovecraftLocation`?

Reply:
Nah, probably somewhere else